lib/blockchain/hardwareWallet.js

```diff
@@ -0,0 +1,305 @@
+      return devices;
+    }
+
+    if (type === '') type = this.type;
+
+    assert(this.isValidDeviceType(type));
+
+    switch (this.type) {
+      case 'ledger':
+        devices.push(...(await Device.getDevices()));
+        break;
+      default:
+        throw new Error(`Unknown device type: ${type}`);
+    }
+
+    this.logger.debug(`found ${devices.length} hardware devices`);
+    return devices;
+  }
+
+  /**
+   * Initialize connection with hardware device
+   * @returns {void}
+   *
+   * TODO: figure out way to allow user to select which device
+   */
+  async initialize() {
+    const unlock = await this.lock.lock();
+
+    const devices = await this.getDevices();
+    const device = devices[0];
+
+    assert(this.initialized === false, 'cannot initialize twice');
+
+    try {
+      switch (this.type) {
+        case 'ledger':
+          this.device = new Device({
+            device: device,
+            timeout: HARDWARE_WALLET_TIMEOUT,
+          });
+          await this.device.open();
+          this.app = new LedgerBcoin({ device: this.device });
+          break;
+
+        default:
+          throw new Error(`Unknown device type: ${this.type}`);
+      }
+
+      this.initialized = true;
+    } catch (e) {
+      this.logger.error(e);
+    } finally {
+      unlock();
+    }
+  }
+
+  /**
+   * Close connection with hardware device
+   * @returns {void}
+   *
+   * TODO: ensure compatibility between devices
+   */
+  async close() {
+    if (this.device) await this.device.close();
+  }
+
+  /**
+   * Close connection with hardware device
+   * and reset state
+   * @returns {void}
+   */
+  async refresh() {
+    this.initialized = false;
+    await this.close();
+
+    this.device = null;
+    this.app = null;
+  }
+
+  /**
+   * Get public key with lock
+   * @param {string} path - bip 39 path
+   * @returns bcoin.HDPublicKey
+   */
+  async getPublicKey(path) {
+    assert(this.initialized, 'must initialize a hardware device');
+    const unlock = await this.lock.lock();
+
+    try {
+      return await this._getPublicKey(path);
+    } catch (e) {
+      this.logger.error(e);
+      this.logger.error(e.stack);
+    } finally {
+      unlock();
+    }
+  }
+
+  /**
+   * Get public key without lock
+   * @private
+   * @param {string} path - bip 39 path
+   * @returns bcoin.HDPublicKey
+   */
+  async _getPublicKey(path) {
+    switch (this.type) {
+      case 'ledger':
+        return await this.app.getPublicKey(path);
+
+      default:
+        return null;
+    }
+  }
+
+  /**
+   * Get signatures for transaction.
+   * @param {bcoin.TX|bcoin.MTX|Buffer} tx - transaction
+   * @param {bcoin.CoinView|Buffer} view
+   * @param {LedgerTXInput[]} ledgerInputs
+   * @returns {Buffer[]}
+   * @throws {LedgerError}
+   * @throws {AssertionError}
+   */
+  async getTransactionSignatures(tx, view, ledgerInputs) {
+    assert(this.initialized, 'must initialize a hardware device');
+    const unlock = await this.lock.lock();
```

Comment:
worth it to add any other assertions around the arguments passed or is it fine to assume that `this.app.getTransactionSignatures` will catch it?